### ISSUE TITLE
Dueling DQN architecture doesn't split dense layer into streams

### ISSUE THREAD

**Issue by Qinusty** [keras-rl/keras-rl#312]
Looking at the implementation of DQNAgent, it appears that the implementation for creating a duelling architecture from the provided model doesn't match the proposed duelling architecture proposed in the [paper](https://arxiv.org/pdf/1511.06581.pdf)

This can be seen in [dqn.py L133](github.com/keras-rl/keras-rl/blob/master/rl/agents/dqn.py#L133), the current implementation uses a single dense layer of size N+1 for N actions but doesn't split the dense layer prior to the output layer as seen below. 

![image](https://user-images.githubusercontent.com/7274845/56815143-7ca29d80-6838-11e9-9b4d-66ad803b966f.png)

Any thoughts on the implications of this would be appreciated.

**Comment by samsenyang** [comment]
I am confused about this issue as well. I saw some implementations in which the output of dense layer was splitted into two streams, such as : https://gist.github.com/awjuliani/35d2ab3409fc818011b6519f0f1629df
https://github.com/p-Mart/Double-Dueling-DQN-Keras/blob/master/DDDQN.py

But I also saw some implementations in whci the outpuf of dense layer was NOT splitted into two streams, such as:
https://github.com/germain-hug/Deep-RL-Keras/blob/master/DDQN/agent.py
https://github.com/MorvanZhou/Reinforcement-learning-with-tensorflow/blob/master/contents/5.3_Dueling_DQN/RL_brain.py

So it is really hard to tell whether we should split it or not. My guess is that maybe it doesn't matter that much if we split it or not.

**Comment by AloshkaD** [comment]
@samsenyang  I also want to figure out weather the streams must be split or not. I think it matters whether you split or not and I think that the correct approach is to split. I think that the sample code for dqn is incorrect.

**Comment by AloshkaD** [comment]
In the [stable_baseline implementation](https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/deepq/policies.py), the network had a V and Q branched from the network.  The used nature_cnn[https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/policies.py](url) does not have a prediction layer which makes it convenient to take it in into the dueling method and branch off the value and prediction. Hence, I firmly believe that the dueling implementation in the keras-rl sample code is incorrect.

**Comment by samsenyang** [comment]
@AloshkaD Yeah. The correct way is defintely splitting the layers into Value layer and Advantage layer. It just confused me that in the experiment of a paper I read, splitting was not used but it seemed that it worked for them. But they still claimed that they used dueling DQN method.

Anyway, I would prefer to split the outputs of previous layer into Value layer and Advantage layer.
Can I ask you on which axis should we split the matrix? In all the examples i have seen, people usually split the matrix on the last axis into two parts, and put each part into Value layer and Advantage layer seperately. Is it a correct way?

**Comment by AloshkaD** [comment]
@samsenyang Agreed.The split will depend on the order of the input layer. In most cases, it's the last axis, just like the example that you shared. Today I'm working on this, and will share the network with you as soon as I test it. My long story short, I've been implementing PPO2 using stable baseline for months, and my simulation isn't learning anything. I used keras-rl previously and DQN was working for me, this is why I'm coming back here.

**Comment by samsenyang** [comment]
@AloshkaD Thank you!

**Comment by AloshkaD** [comment]
@samsenyang on a second thought, and upon looking at the dqn code, I think that the dqn does take the layer before the last one (prediction) and splits it into A and V

```class DQNAgent(AbstractDQNAgent):
    """
    # Arguments 
        model__: A Keras model. 
        policy__: A Keras-rl policy that are defined in [policy](https://github.com/keras-rl/keras-rl/blob/master/rl/policy.py). 
        test_policy__: A Keras-rl policy. 
        enable_double_dqn__: A boolean which enable target network as a second network proposed by van Hasselt et al. to decrease overfitting. 
        enable_dueling_dqn__: A boolean which enable dueling architecture proposed by Mnih et al. 
        dueling_type__: If `enable_dueling_dqn` is set to `True`, a type of dueling architecture must be chosen which calculate Q(s,a) from V(s) and A(s,a) differently. Note that `avg` is recommanded in the [paper](https://arxiv.org/abs/1511.06581). 
            `avg`: Q(s,a;theta) = V(s;theta) + (A(s,a;theta)-Avg_a(A(s,a;theta))) 
            `max`: Q(s,a;theta) = V(s;theta) + (A(s,a;theta)-max_a(A(s,a;theta))) 
            `naive`: Q(s,a;theta) = V(s;theta) + A(s,a;theta) 
 
    """
    def __init__(self, model, policy=None, test_policy=None, enable_double_dqn=True, enable_dueling_network=False,
                 dueling_type='avg', *args, **kwargs):
        super(DQNAgent, self).__init__(*args, **kwargs)

        # Validate (important) input.
        if hasattr(model.output, '__len__') and len(model.output) > 1:
            raise ValueError('Model "{}" has more than one output. DQN expects a model that has a single output.'.format(model))
        if model.output._keras_shape != (None, self.nb_actions):
            raise ValueError('Model output "{}" has invalid shape. DQN expects a model that has one dimension for each action, in this case {}.'.format(model.output, self.nb_actions))

        # Parameters.
        self.enable_double_dqn = enable_double_dqn
        self.enable_dueling_network = enable_dueling_network
        self.dueling_type = dueling_type
        if self.enable_dueling_network:
            # get the second last layer of the model, abandon the last layer
            layer = model.layers[-2]
            nb_action = model.output._keras_shape[-1]
            # layer y has a shape (nb_action+1,)
            # y[:,0] represents V(s;theta)
            # y[:,1:] represents A(s,a;theta)
            y = Dense(nb_action + 1, activation='linear')(layer.output)
            # caculate the Q(s,a;theta)
            # dueling_type == 'avg'
            # Q(s,a;theta) = V(s;theta) + (A(s,a;theta)-Avg_a(A(s,a;theta)))
            # dueling_type == 'max'
            # Q(s,a;theta) = V(s;theta) + (A(s,a;theta)-max_a(A(s,a;theta)))
            # dueling_type == 'naive'
            # Q(s,a;theta) = V(s;theta) + A(s,a;theta)
            if self.dueling_type == 'avg':
                outputlayer = Lambda(lambda a: K.expand_dims(a[:, 0], -1) + a[:, 1:] - K.mean(a[:, 1:], keepdims=True), output_shape=(nb_action,))(y)
            elif self.dueling_type == 'max':
                outputlayer = Lambda(lambda a: K.expand_dims(a[:, 0], -1) + a[:, 1:] - K.max(a[:, 1:], keepdims=True), output_shape=(nb_action,))(y)
            elif self.dueling_type == 'naive':
                outputlayer = Lambda(lambda a: K.expand_dims(a[:, 0], -1) + a[:, 1:], output_shape=(nb_action,))(y)
            else:
                assert False, "dueling_type must be one of {'avg','max','naive'}"

            model = Model(inputs=model.input, outputs=outputlayer)```
